File: LinkedList/1721.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        slow, fast, dummy = head, head, head
        num_1, num_2 = 0, 0
        tmp_k = k
        
        while k - 1 != 0:
            fast = fast.next
            k -= 1
        num_1 = fast.val
        
        while fast.next:
            slow = slow.next
            fast = fast.next
        
        num_2 = slow.val
        slow.val = num_1
        
        while dummy:
            tmp_k -= 1
            if tmp_k == 0:
                dummy.val = num_2
                break
            dummy = dummy.next
        return head

File: LinkedList/1721.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {        
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* headPtr = head;
        int count = k;
        int temp1, temp2;
        while (k > 1) {
            fast = fast->next;
            k--;
        }
        temp1 = fast->val;        
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next;
        }
        temp2 = slow->val;        
        slow->val = temp1;
        while (count > 1) {
            headPtr = headPtr->next;
            count--;
        }
        headPtr->val = temp2;
        return head;
    }
};
